base/dcp_dest.go

```diff
@@ -0,0 +1,406 @@
+}
+
+// DCPLoggingDest wraps DCPDest to provide per-callback logging
+type DCPLoggingDest struct {
+	dest *DCPDest
+}
+
+func (d *DCPLoggingDest) Close() error {
+	return d.dest.Close()
+}
+
+func (d *DCPLoggingDest) DataUpdate(partition string, key []byte, seq uint64,
+	val []byte, cas uint64, extrasType cbgt.DestExtrasType, extras []byte) error {
+
+	TracefCtx(d.dest.loggingCtx, KeyDCP, "DataUpdate:%s, %s, %d", partition, UD(string(key)), seq)
+	return d.dest.DataUpdate(partition, key, seq, val, cas, extrasType, extras)
+}
+
+func (d *DCPLoggingDest) DataUpdateEx(partition string, key []byte, seq uint64, val []byte,
+	cas uint64, extrasType cbgt.DestExtrasType, req interface{}) error {
+
+	TracefCtx(d.dest.loggingCtx, KeyDCP, "DataUpdateEx:%s, %s, %d", partition, UD(string(key)), seq)
+	return d.dest.DataUpdateEx(partition, key, seq, val, cas, extrasType, req)
+}
+
+func (d *DCPLoggingDest) DataDelete(partition string, key []byte, seq uint64,
+	cas uint64, extrasType cbgt.DestExtrasType, extras []byte) error {
+	TracefCtx(d.dest.loggingCtx, KeyDCP, "DataDelete:%s, %s, %d", partition, UD(string(key)), seq)
+	return d.dest.DataDelete(partition, key, seq, cas, extrasType, extras)
+}
+
+func (d *DCPLoggingDest) DataDeleteEx(partition string, key []byte, seq uint64,
+	cas uint64, extrasType cbgt.DestExtrasType, req interface{}) error {
+	TracefCtx(d.dest.loggingCtx, KeyDCP, "DataDeleteEx:%s, %s, %d", partition, UD(string(key)), seq)
+	return d.dest.DataDeleteEx(partition, key, seq, cas, extrasType, req)
+}
+
+func (d *DCPLoggingDest) SnapshotStart(partition string,
+	snapStart, snapEnd uint64) error {
+	TracefCtx(d.dest.loggingCtx, KeyDCP, "SnapshotStart:%d, %d, %d", partition, snapStart, snapEnd)
+	return d.dest.SnapshotStart(partition, snapStart, snapEnd)
+}
+
+func (d *DCPLoggingDest) OpaqueGet(partition string) (value []byte, lastSeq uint64, err error) {
+	TracefCtx(d.dest.loggingCtx, KeyDCP, "OpaqueGet:%s", partition)
+	return d.dest.OpaqueGet(partition)
+}
+
+func (d *DCPLoggingDest) OpaqueSet(partition string, value []byte) error {
+	TracefCtx(d.dest.loggingCtx, KeyDCP, "OpaqueSet:%s, %s", partition, value)
+	return d.dest.OpaqueSet(partition, value)
+}
+
+func (d *DCPLoggingDest) Rollback(partition string, rollbackSeq uint64) error {
+	InfofCtx(d.dest.loggingCtx, KeyDCP, "Rollback:%s, %d", partition, rollbackSeq)
+	return d.dest.Rollback(partition, rollbackSeq)
+}
+
+func (d *DCPLoggingDest) RollbackEx(partition string, vbucketUUID uint64, rollbackSeq uint64) error {
+	InfofCtx(d.dest.loggingCtx, KeyDCP, "RollbackEx:%s, %v, %d", partition, vbucketUUID, rollbackSeq)
+	return d.dest.RollbackEx(partition, vbucketUUID, rollbackSeq)
+}
+
+func (d *DCPLoggingDest) ConsistencyWait(partition, partitionUUID string,
+	consistencyLevel string, consistencySeq uint64, cancelCh <-chan bool) error {
+	return d.dest.ConsistencyWait(partition, partitionUUID, consistencyLevel, consistencySeq, cancelCh)
```

Comment:
Should we add the logging even for the no-op functions? Just in-case we need to add an implementation for them in the future.

```suggestion
 	InfofCtx(d.dest.loggingCtx, KeyDCP, "ConsistencyWait:%s, %s, %s, %d", partition, partitionUUID, consistencyLevel, consistencySeq) 
	return d.dest.ConsistencyWait(partition, partitionUUID, consistencyLevel, consistencySeq, cancelCh)
```